## A Numerical Solver for the Dual Variance Problem

### Review of the Problem

This Jupyter notebook accompanies Section 3.6 of my PhD Thesis, and provides a method to explicitly compute the dual supremum. 
In particular, we use the formulas in Proposition 3.19 and Corollary 3.20 to reduce this computation to a finite-dimensional optimization problem over $(\lambda,\eta)\in\R^2\times (-\infty,0).$

First, import some libraries:

In [399]:
import plotly.graph_objects as go
import numpy as np
import scipy.optimize as opt
import warnings
warnings.filterwarnings("ignore", category=np.ComplexWarning)

Next, we implement Equation (3.15) in Proposition 3.19 of my thesis which, given $x\in [0,\infty)\times (0,\infty)$ and $(\lambda,\eta)\in \R^2\times (-\infty,0)$, allows us to compute $$Y(x;\lambda,\eta)\in \operatorname{argmin}_{y\in[0,\infty)^2} [c(x,y)-\varphi(x)-\lambda\cdot y-\eta|y-m|^2].$$ 

In [ ]:
# Start by defining a few auxiliary functions
def func_A(x2: float, eta : float) -> float:
    return - 4 *  x2 * eta

def func_B1(x1 : float, x2: float, lda1 : float, eta : float, m1: float) -> float:
    return 2 * x1 + 2 * x2 * lda1 - 4 * x2 * eta * m1

def func_B2(x1 : float, x2: float, lda2 : float, eta : float,  m2 : float) -> float:
    return x1**2 - 2 * x2 * lda2 + 4 * x2 * eta * m2

def func_C(x1 : float, lda1 : float, lda2 : float, eta : float, m1: float, m2 : float) -> float:
    return (2 * eta * (m1 * x1 + m2) - (lda1 * x1 +lda2))/(2* eta * (x1**2 +1))

# Next, define the cost function
def func_c(x1: float, x2: float, y1: float, y2: float) -> float:
    if (x2 > 0) and (y2 > 0):
        return (y2/(2*x2))*(x1-y1/y2)**2
    elif x1 * y2 - y1 == 0:
        return 0
    else: 
        return float('inf') 

# Define the largest solution of $g(u)=u^3+(2-B_2)u-B_1=0$
def func_u0(B1: float, B2: float) -> float:
    return np.max(np.roots([1, 0, 2-B2, -B1]))

# Compute $Y$ as in Proposition 3.19.
def func_Y(x1 : float, x2: float, lda1 : float, lda2 : float, eta : float, m1: float, m2 : float) -> np.ndarray:
    A = func_A(x2,eta)
    B1 = func_B1(x1, x2, lda1, eta, m1)
    B2 = func_B2(x1, x2, lda2, eta, m2)
    C = func_C(x1, lda1, lda2, eta, m1, m2)
    u0 = func_u0(B1,B2)
    if (x2 > 0) and (B1 > 0) and (4*B2 < B1**2):
        return ((u0**2 - B2)/A) * np.array([u0, 1])
    elif (x2 > 0) and (B1 <= 0) and (B2 < 0):
        return np.array([0,-B2/A])
    elif (x2 == 0) and (C > 0):
        return C*np.array([x1,1])
    else:
        return np.array([0,0])
    
# Compute the measure Y_\#\mu.
def compute_pushforward(mu, lda, eta, m = mean):
    nu = mu.copy()
    results = np.array([func_Y(x1, x2, lda[0], lda[1], eta, m[0], m[1]) for x1, x2 in zip(nu[:,1], nu[:,2])])
    nu[:,1] = results[:,0]
    nu[:,2] = results[:,1]
    return nu
    

Next we define some graphing functions using the `plotly` Python library:

In [401]:
def plot_atomic_measure_3d(mu, mean = None, params = None, variable = "x"):
    """
    Plot a discrete measure of the form
        mu[i] = [weight_i, x_1,i, x_2,i]
    as a 3D scatter plot where height = weight.

    Parameters
    ----------
    mu : (n,3) numpy array
        First column = weights
        Second column = x_1-coordinates
        Third column = x_2-coordinates

    mean : (1,2) numpy array
        First  column = x_1-coordinates
        Second column = x_2-coordinates
        If provided, plots a vertical line at x1/x2 ratio and a label
    """

    mu = np.asarray(mu, dtype=float)
    if mu.ndim != 2 or mu.shape[1] != 3:
        raise ValueError("mu must be an (n,3) array: [weight, x_1, x_2].")

    weights = mu[:, 0]
    x1 = mu[:, 1]
    x2 = mu[:, 2]

    # --- Scatter points ---
    if variable == "x":
        scatter = go.Scatter3d(
            x=x1,
            y=x2,
            z=weights,
            mode='markers',
            marker=dict(size=5, color='blue'),
            name='&#956;'
        )
    if variable == 'y':
        scatter = go.Scatter3d(
            x=x1,
            y=x2,
            z=weights,
            mode='markers',
            marker=dict(size=5, color='blue'),
            name='Y<sub>&#35;</sub>&#956;'
        )

    # --- Mass pillars ---
    pillar_lines = []
    for xi, yi, wi in zip(x1, x2, weights):
        pillar_lines.append(
            go.Scatter3d(
                x=[xi, xi],
                y=[yi, yi],
                z=[0, wi],
                mode='lines',
                line=dict(color='blue', width=4),
                showlegend=False
            )
        )

    # --- Vertical line at mean (optional) ---
    mean_line = None
    mean_label = None
    point = None
    if mean is not None:
        if variable == 'x':
            m = mean[0]/mean[1]
            y_min, y_max = x2.min()-0.2, x2.max()+0.2
            mean_line = go.Scatter3d(
                x=[m, m],
                y=[y_min, y_max],
                z=[0, 0],
                mode='lines',
                line=dict(color='red', width=3, dash='dash'),
                name='x<sub>1</sub>=m<sub>1</sub>/m<sub>2</sub>'
            )
            # label at midpoint
            y_mid = (y_min + y_max)/2
            mean_label = go.Scatter3d(
                x=[m],
                y=[y_mid],
                z=[0.01],  # tiny offset above plane
                mode='text',
                text=["x<sub>1</sub>=m<sub>1</sub>/m<sub>2</sub>"],
                textposition="middle right",
                textfont=dict(color='red', size=14),
                showlegend=False
            )
        if variable == 'y':
            point = go.Scatter3d(
                    x=[mean[0]],
                    y=[mean[1]],
                    z=[0],
                    mode='markers+text',
                    marker=dict(color='red', size=5),
                    text=["(m<sub>1</sub>, m<sub>2</sub>)"],
                    textposition="middle right",
                    textfont=dict(color='red', size=14),
                    name='(m<sub>1</sub>, m<sub>2</sub>)'
                )

    # --- Compose all traces ---
    traces = [scatter] + pillar_lines
    if mean_line:
        traces.append(mean_line)
    if mean_label:
        traces.append(mean_label)
    if point:
        traces.append(point)

    # --- Layout ---
    if variable == 'x':
        layout = go.Layout(
            scene=dict(
                xaxis_title=variable + "<sub>1</sub> (emissions intensity)",
                yaxis_title=variable + "<sub>2</sub> (ease of change)",
                zaxis_title='Weight',
            ),
            #title=('Support of the Post-regulation Distribution Y<sub>&#35;</sub>&#956; with &#955; = ('
            #           + str(params[0].round(4)) + ", " + str(params[1].round(4))
            #        + ") and &#951; = " +str(-np.exp(params[2]).round(4))),
            title=('The Pre-regulation Distribution &#956;'),
            margin=dict(l=0, r=0, b=0, t=60),
            showlegend=True
        )
    if variable == 'y':
        layout = go.Layout(
            scene=dict(
                xaxis_title=variable + "<sub>1</sub> (emissions)",
                yaxis_title=variable + "<sub>2</sub> (production)",
                zaxis_title='Weight',
            ),
            title=('The Post-regulation Distribution Y<sub>&#35;</sub>&#956; with &#955; = ('
                       + str(params[0].round(4)) + ", " + str(params[1].round(4))
                    + ") and &#951; = " +str(-np.exp(params[2]).round(4))),
            margin=dict(l=0, r=0, b=0, t=60),
            showlegend=True
        )
    fig = go.Figure(data=traces, layout=layout)
    return fig
def plot_atomic_measure_2d(mu, mean=None, params = None, variable="x"):
    """
    Plot a discrete measure of the form
        mu[i] = [weight_i, x_1,i, x_2,i]
    as a 2D scatter plot (weights ignored).

    Parameters
    ----------
    mu : (n,3) numpy array
        First column = weights (ignored)
        Second column = x_1-coordinates
        Third column = x_2-coordinates

    mean : (1,2) numpy array
        First  column = x_1-coordinates
        Second column = x_2-coordinates
        If provided, plots a vertical line at x1/x2 ratio and a label
    """
    mu = np.asarray(mu, dtype=float)
    if mu.ndim != 2 or mu.shape[1] != 3:
        raise ValueError("mu must be an (n,3) array: [weight, x_1, x_2].")

    x1 = mu[:, 1]
    x2 = mu[:, 2]

    # --- Scatter points ---
    if variable == 'x':
        traces = [go.Scatter(
            x=x1,
            y=x2,
            mode='markers',
            marker=dict(size=8, color='blue'),
            name='spt(&#956;)'
        )]
    if variable == 'y':
        traces = [go.Scatter(
            x=x1,
            y=x2,
            mode='markers',
            marker=dict(size=8, color='blue'),
            name='spt(Y<sub>&#35;</sub>&#956;)'
        )]

    # --- Mean overlay (optional) ---
    if mean is not None:
        if variable == 'x':
            m = mean[0] / mean[1]
            x_min, x_max = min(x1.min(),m) - 0.2, max(x1.max(),m) + 0.2
            y_min, y_max = x2.min() - 0.2, x2.max() + 0.2
            traces.append(go.Scatter(
                x=[m, m],
                y=[y_min, y_max],
                mode='lines',
                line=dict(color='red', width=2, dash='dash'),
                name='x<sub>1</sub>=m<sub>1</sub>/m<sub>2</sub>'
            ))
        elif variable == 'y':
            traces.append(go.Scatter(
                x=[mean[0]],
                y=[mean[1]],
                mode='markers+text',
                marker=dict(color='red', size=8),
                text=["(m<sub>1</sub>, m<sub>2</sub>)"],
                textposition="middle right",
                textfont=dict(color='red', size=14),
                name='(m<sub>1</sub>, m<sub>2</sub>)'
            ))

    # --- Layout ---
    if variable == 'x':
        fig = go.Figure(data=traces, layout=go.Layout(
            xaxis = dict(title = variable + "<sub>1</sub> (emissions intensity)", range = [x_min,x_max]),
            yaxis_title=variable + "<sub>2</sub> (ease of change)",
            title=('Support of the Pre-regulation Distribution &#956;'),
            margin=dict(l=0, r=0, b=0, t=60),
            showlegend=True
        )) 
    if variable == 'y':
        fig = go.Figure(data=traces, layout=go.Layout(
            xaxis_title=variable + "<sub>1</sub> (emissions)",
            yaxis_title=variable + "<sub>2</sub> (production)",
            title=('Support of the Post-regulation Distribution Y<sub>&#35;</sub>&#956; with &#955; = ('
                   + str(params[0].round(4)) + ", " + str(params[1].round(4))
                + ") and &#951; = " +str(-np.exp(params[2]).round(4))),
            margin=dict(l=0, r=0, b=0, t=60),
            showlegend=True
        ))

    return fig

def plot_marginal_comparison(mu, nu, params, bins=100):
    """
    Plot the x1 distribution of mu against the y1/y2 distribution of nu.

    Parameters
    ----------
    mu : (n,3) array
        Columns: [weight, x1, x2]
    nu : (m,3) array
        Columns: [weight, y1, y2]
    bins : int
        Number of histogram bins
    """
    mu = np.asarray(mu, dtype=float)
    nu = np.asarray(nu, dtype=float)

    # x1 marginal from mu
    x1 = mu[:,1]
    w_mu = mu[:,0]/mu[:,0].sum()  # normalize to relative weight

    # y1/y2 marginal from nu
    y_ratio = nu[:,1]/nu[:,2]  # y1/y2 ratio
    w_nu = nu[:,0]/nu[:,0].sum()


    # Compute shared bin edges spanning both distributions
    shared_min = min(x1.min(), y_ratio.min())
    shared_max = max(x1.max(), y_ratio.max())
    edges = np.linspace(shared_min-0.1, shared_max+0.1, bins + 1)

    hist_mu, _ = np.histogram(x1,      bins=edges, weights=w_mu, density=True)
    hist_nu, _ = np.histogram(y_ratio, bins=edges, weights=w_nu, density=True)

    centers = (edges[:-1] + edges[1:]) / 2
    width   = (edges[1] - edges[0]) * 0.9

    # Plot
    fig = go.Figure()
    fig.add_trace(go.Bar(
        x=centers,
        y=hist_mu,
        width=width,
        name='Pre-reg. distribution x<sub>1</sub>',
        marker_color='blue',
        opacity=0.6
    ))
    fig.add_trace(go.Bar(
        x=centers,
        y=hist_nu,
        width=width,
        name='Post-reg. distribution y<sub>1</sub>/y<sub>2</sub>',
        marker_color='red',
        opacity=0.6
    ))

    fig.update_layout(
        barmode='overlay',
        xaxis_title='Emissions Intensity',
        yaxis_title='Percentage of Firms',
        title=('Emissions Intensity Distributions with &#955; = ('+ str(params[0].round(4)) + ", " + str(params[1].round(4))
               + ") and &#951; = " +str(-np.exp(params[2]).round(4))),
        legend=dict(x=0.7, y=0.95),
        xaxis=dict(dtick=0.1)
    )

    return fig

Now, we define a measure and compute key quantities. We encode discrete probability measures as `numpy` `ndarrays`, where each row has three entries -- the mass, the $x_1$ coordinate, and the $x_2$ coordinate. We also set a mean target and a variance target.

In [402]:
# Define one sample measure for which the dual supremum appears to be attained
mu_denorm_attain = np.array([[1, 1, 1], [1, 2, 1], [1, 1, 2], [1, 2,2]], dtype = 'f')
# Define another sample measure for which the dual supremum appears not to be attained. 
mu_denorm_nonattain = np.array([[1, 2, 4]], dtype = 'f')
# Automatically normalize both measures
mu_norm_attain = mu_denorm_attain
mu_norm_attain[:,0] = mu_norm_attain[:,0]/(np.sum(mu_norm_attain[:,0], axis = 0))
mu_norm_nonattain = mu_denorm_nonattain
mu_norm_nonattain[:,0] = mu_norm_nonattain[:,0]/(np.sum(mu_norm_nonattain[:,0], axis = 0))

# Set mean target
mean = np.array([1,2])
# Set variance target
tau = 1.5

In [403]:
plot_atomic_measure_3d(mu_norm_attain, mean = mean).write_html("prereg.html")
plot_atomic_measure_3d(mu_norm_attain, mean = mean)

In [404]:
plot_atomic_measure_2d(mu_norm_attain, mean = mean)

The choice of $Y=[0,\infty)^2,$ along with the presence of a prescribed mean, allows us to compute an optimal value of $\varphi$ as in Corollary 3.20:


In [405]:
def optimal_phi(x1: float, x2: float, lda1: float, lda2: float, eta: float) -> float:
    y1,y2 = func_Y(x1, x2, lda1, lda2, eta, mean[0], mean[1])
    return func_c(x1, x2, y1, y2) - lda1 * y1 - lda2 * y2 - eta * ((y1-mean[0])**2 + (y2-mean[1])**2) 


Next, we define a function to compute the dual objective function 
$$\int \varphi\; d\mu+\lambda\cdot y+\eta\tau.$$

In [406]:
def objective_function(params, measure) -> float:
    lda1, lda2, theta = params
    eta = -np.exp(theta) # technical tweak to get an unbounded domain
    # negative because scipy can only minimize
    return (-sum([measure[i,0] * optimal_phi(measure[i, 1], measure[i, 2], lda1, lda2, eta) for i in range(len(measure))])
            - lda1 * mean[0] - lda2 * mean[1] - eta * tau )

We will also wish to compute the quantities $$\int Y(\cdot;\lambda,\eta)\; d\mu$$ and $$\int |Y(\cdot;\lambda,\eta)-m|^2\; d\mu$$ to verify if our result actually achieves the mean and variance constraints:

In [407]:
def mu_mean(params, measure, m= mean) -> np.ndarray:
    lda1, lda2, theta = params
    eta = -np.exp(theta)
    return sum([measure[i,0] * func_Y(measure[i,1], measure[i,2], lda1, lda2, eta, m[0], m[1]) for i in range(len(measure))])

def mu_variance(params, measure) -> np.ndarray:
    lda1, lda2, theta = params
    eta = -np.exp(theta)
    return sum([measure[i,0] * np.linalg.norm(func_Y(measure[i,1], measure[i,2], lda1, lda2, eta, mean[0], mean[1])-mean)**2 
                for i in range(len(measure))])

Moreover, we compute $$\int c(x, Y(x;\lambda,\eta))d\mu(x)$$ in order to get a sense of the value of the primal infimum.

In [408]:
def primal_cost(params, measure) -> np.ndarray:
    lda1, lda2, theta = params
    eta = -np.exp(theta)
    return sum([measure[i,0] * func_c(measure[i,1], measure[i,2], 
                                      func_Y(measure[i,1], measure[i,2], lda1, lda2, eta, mean[0], mean[1])[0],
                                      func_Y(measure[i,1], measure[i,2], lda1, lda2, eta, mean[0], mean[1])[1]) 
                                      for i in range(len(measure))])

Finally, for technical reasons, we also create a wrapper so that we can use SciPy's optimize function. 

In [409]:
def make_objective(measure):
    def wrapped_objective(params):
        return objective_function(params, measure)
    return wrapped_objective

Next, we finally complete the optimization:

In [410]:
optimization_result_attain = opt.minimize(make_objective(mu_norm_attain), [-1,1,-1], method = 'Nelder-Mead')

We recover the optimal choices of $(\lambda,\eta).$ 

In [411]:
optimal_params_attain = optimization_result_attain.x
print("lambda = " +  str(optimal_params_attain[0:2]))
print("eta = " +  str(-np.exp(optimal_params_attain[2])))

lambda = [-0.47282743  0.54774745]
eta = -0.09954221634326921


Next we use this to recover the target measure:

In [412]:
nu = compute_pushforward(mu_norm_attain, optimal_params_attain[0:2], -np.exp(optimal_params_attain[2]), mean)

In [413]:
nu

array([[0.25      , 1.3019906 , 2.7876847 ],
       [0.25      , 0.9594234 , 0.6249295 ],
       [0.25      , 0.66481197, 3.5398757 ],
       [0.25      , 1.0737611 , 1.0475965 ]], dtype=float32)

In [414]:
plot_atomic_measure_3d(nu, mean = mean, params = optimal_params_attain, variable = 'y').write_html('postreg.html')
plot_atomic_measure_3d(nu, mean = mean, params = optimal_params_attain, variable = 'y')

In [415]:
plot_atomic_measure_2d(nu, mean = mean, params = optimal_params_attain, variable = 'y')

In [416]:
plot_marginal_comparison(mu_norm_attain, nu, optimal_params_attain, bins = 100)

Next, we verify that the result achieves the target mean and variance, which is the case for `mu_norm_attain`. We also compute the value of the objective function, which in this case should be approximately equal to the dual supremum. 

In [417]:
print("mean: " +  str(mu_mean(optimal_params_attain, mu_norm_attain)))
print("variance: " + str(mu_variance(optimal_params_attain, mu_norm_attain)))
print("dual value: " + str(-objective_function(optimal_params_attain, mu_norm_attain)))
print("primal value: " + str(primal_cost(optimal_params_attain, mu_norm_attain)))

mean: [0.99999678+0.j 2.00002156+0.j]
variance: 1.5000480053980814
dual value: (0.32402810105060625+0j)
primal value: (0.3240366522395013+0j)


In [418]:
nu = compute_pushforward(mu_norm_attain, optimal_params_attain[0:2], -np.exp(optimal_params_attain[2]), mean)
np.sum(nu[:,0:1] * nu[:,1:3], axis=0)

array([0.9999968, 2.0000215], dtype=float32)

On the other hand, `mu_norm_nonattain` does not meet the target mean and variance, indicating non-attainment of the dual supremum (which we have already seen in Proposition 3.1):

In [419]:
optimization_result_nonattain = opt.minimize(make_objective(mu_norm_nonattain), [-1,1,-1], method = 'Nelder-Mead')
optimal_params_nonattain = optimization_result_nonattain.x
optimal_params_nonattain[2] = -np.exp(optimal_params_nonattain[2])
print("lambda = " +  str(optimal_params_nonattain[0:2]))
print("eta = " +  str(optimal_params_nonattain[2]))
print("mean: " +  str(mu_mean(optimal_params_nonattain, mu_norm_nonattain)))
print("variance: " + str(mu_variance(optimal_params_nonattain, mu_norm_nonattain)))
print("dual value: " + str(-objective_function(optimal_params_nonattain, mu_norm_nonattain)))
print("primal value: " + str(primal_cost(optimal_params_nonattain, mu_norm_nonattain)))


lambda = [-0.375    0.46875]
eta = -1.7360330996572665e-16
mean: [1.+0.j 2.+0.j]
variance: 3.836616882424559e-18
dual value: (-0.9374999999999996+0j)
primal value: (0.5624999989324116+0j)


Observe that the optimization algorithm yields $\eta \approx -1.7\times 10^{-16}\approx 0.$ We interpret this to mean that the optimization procedure approaches $\eta=0,$ but discontinuity of the objective function here prevents the existence of a maximum. 